# Exercises Part I

In [1]:
import pandas as pd
import numpy as np
from pydataset import data
from env import host, user, password

In [52]:
def get_db_url(user,host,password):
    url = f'mysql+pymysql://{user}:{password}@{host}/employees'
    return url

In [3]:
url = get_db_url(user,host,password)

In [4]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)


emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

## Q5

Q5a

NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:ysql.pymysql

Q5b

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FOM employees LIMIT 5 OFFSET 50' at line 1")
[SQL: SELECT * FOM employees LIMIT 5 OFFSET 50]
(Background on this error at: https://sqlalche.me/e/14/f405)

## Q6

In [5]:
employees_df = pd.read_sql('SELECT * FROM employees', url)

In [6]:
employees_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [7]:
titles_df = pd.read_sql('SELECT * FROM titles', url)

In [8]:
titles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


## Q7

#### employees, 300024 rows × 6 columns
#### titles,    443308 rows × 4 columns

This is expected because the employees data base is very large and tables differ in columns as tables are often referenced to one another in the database.

## Q8

In [9]:
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [10]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

## Q9

In [11]:
pd.read_sql('SELECT DISTINCT title FROM titles', url)

title
0     Senior Engineer
1               Staff
2            Engineer
3        Senior Staff
4  Assistant Engineer
5    Technique Leader
6             Manager

## Q10

In [12]:
pd.read_sql('SELECT to_date FROM titles ORDER BY to_date LIMIT 10', url)

to_date
0  1985-03-01
1  1985-03-11
2  1985-03-20
3  1985-03-23
4  1985-04-17
5  1985-04-20
6  1985-04-21
7  1985-04-23
8  1985-05-05
9  1985-05-05

Oldest to_date is 1985-03-01

## Q11

In [13]:
pd.read_sql('SELECT to_date FROM titles ORDER BY to_date DESC LIMIT 10', url)

to_date
0  9999-01-01
1  9999-01-01
2  9999-01-01
3  9999-01-01
4  9999-01-01
5  9999-01-01
6  9999-01-01
7  9999-01-01
8  9999-01-01
9  9999-01-01

Most recent to_date is 9999-01-01

# Exercises Part II

## Q1

In [14]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [15]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

## Q2

In [16]:
users.merge(roles, how='right')

id       name  role_id
0   1      admin      NaN
1   2     author      NaN
2   3   reviewer      NaN
3   4  commenter      NaN

## Q3

In [17]:
users.merge(roles, how='left')

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

## Q4

In [18]:
usersdropid = users.drop(columns='id')

In [19]:
usersdropid.merge(roles, how='left')

name  role_id  id
0    bob      1.0 NaN
1    joe      2.0 NaN
2  sally      3.0 NaN
3   adam      3.0 NaN
4   jane      NaN NaN
5   mike      NaN NaN

Missing foreign key generates many null values in resulting table

## Q5, Q6, Q7

In [20]:
mpg = data('mpg')

In [21]:
mpg.shape

(234, 11)

234 rows and 11 columns

## Q9

In [22]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

In [23]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


## Q10

In [24]:
mpg.manufacturer.nunique()

15

## Q11

In [25]:
mpg.model.nunique()

38

## Q12

In [26]:
mpg['mileage_difference'] = mpg.hwy - mpg.cty
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

In [27]:
mpg['average_mileage']= round(2 / ((1/mpg.hwy) + (1/mpg.cty)), 2)

In [28]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   mileage_difference  average_mileage  
1                  11            22.21  
2                   8            24.36  
3                  11            24.31  
4                   9            24.71  
5                  10            19.81

In [29]:
#mpg = mpg.assign(is_automatic= mpg['trans'].str.contains('auto'))

In [30]:
mpg['is_automatic'] = mpg.trans.str.contains('auto')

In [31]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   mileage_difference  average_mileage  is_automatic  
1                  11            22.21          True  
2                   8            24.36         False  
3                  11            24.31         False  
4                   9            24.71          True  
5                  10            19.81          True

## Q15

In [32]:
#mpg.sort_values(by=['manufacturer','average_mileage'], ascending=False)

In [33]:
mpg.groupby('manufacturer').average_mileage.mean().nlargest(3)

manufacturer
honda         27.891111
volkswagen    24.370000
hyundai       22.002143
Name: average_mileage, dtype: float64

Honda has the better average_mileage

## Q16

In [34]:
#mpg.sort_values(by='is_automatic', ascending=False)

In [35]:
#mpg[mpg['is_automatic']==True].sort_values(by='average_mileage', ascending=False)

In [36]:
mpg['trans_category']=np.where(mpg.trans.str.contains('auto'),'automatic','manual')

In [37]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   mileage_difference  average_mileage  is_automatic trans_category  
1                  11            22.21          True      automatic  
2                   8            24.36         False         manual  
3                  11            24.31         False         manual  
4                   9            24.71          True      automatic  
5                  10            19.81          True      automatic

In [38]:
mpg.groupby('trans_category')[['average_mileage']].mean().round(1)

average_mileage
trans_category                 
automatic                  18.6
manual                     21.6

In [39]:
#mpg[mpg['is_automatic']==False].sort_values(by='average_mileage', ascending=False)

Manual transmission cars have better mileage

# Exercises Part III

## Q1

In [141]:
def get_db_url(user,host,password):
    url = f'mysql+pymysql://{user}:{password}@{host}/chipotle'
    return url

In [142]:
url = get_db_url(user,host,password)

In [143]:
#pd.read_sql('SELECT * FROM orders', url)

In [91]:
chipotle = pd.read_sql('SELECT * FROM orders', url)

In [127]:
chipotle['item_price'] = chipotle['item_price'].str.replace('$','').astype('float').round(2)

/var/folders/v_/4rm6f1gd4zsc7jhj61h61kw80000gn/T/ipykernel_18900/553435113.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  chipotle['item_price'] = chipotle['item_price'].str.replace('$','').astype('float').round(2)


In [135]:
chipotle = chipotle.drop(columns='float_price')

In [136]:
chipotle.groupby('order_id').item_price.sum()

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64

In [137]:
chipotle.groupby('item_name').quantity.sum().nlargest(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

In [138]:
chipotle.groupby('item_name').item_price.sum().nlargest()

item_name
Chicken Bowl           7342.73
Chicken Burrito        5575.82
Steak Burrito          3851.43
Steak Bowl             2260.19
Chips and Guacamole    2201.04
Name: item_price, dtype: float64

In [140]:
#chipotle.head(10)

## Q5

In [144]:
def get_db_url(user,host,password,dbname):
    url = f'mysql+pymysql://{user}:{password}@{host}/{dbname}'
    return url

In [148]:
url = get_db_url(user,host,password,'employees')

In [152]:
emp_titles = pd.read_sql('SELECT * FROM employees LEFT JOIN titles USING(emp_no) LIMIT 10', url)